backend/src/main/java/com/empire/World.java

```diff
@@ -1660,19 +1682,44 @@ void gainIncome() {
 		}
 
 		void churchOpinionChangesDueToStateReligion() {
+			Map<String, Ideology> stateIdeologies = kingdoms.keySet().stream().collect(Collectors.toMap(k -> k, k -> NationData.getStateReligion(k, World.this)));
 			for (String k : kingdoms.keySet()) {
-				if (NationData.getStateReligion(k, World.this).religion != Religion.IRUHAN) getNation(k).goodwill -= 5;
-				if (getNation(k).loyalToCult) getNation(k).goodwill -= 10;
+				Religion religion = stateIdeologies.get(k).religion;
+				if (church.hasDoctrine(Church.Doctrine.ANTIECUMENISM) && religion != Religion.IRUHAN && religion != Religion.NONE) getNation(k).goodwill += Constants.antiecumenismStateOpinion;
+				if (church.hasDoctrine(Church.Doctrine.ANTIAPOSTASY) && getNation(k).loyalToCult) getNation(k).goodwill += Constants.antiapostasyOpinion;
+			}
+			Set<String> iruhanNations = stateIdeologies.entrySet().stream().filter(e -> e.getValue().religion == Religion.IRUHAN).map(e -> e.getKey()).collect(Collectors.toSet());
```

Comment:
Could replace these `e -> e.getKey()` with their functional reference `Map.Entry::getKey` if that's your style.

Reply:
Done.

Reply:
Done.